src/neuron/container/generic_data_handle.hpp

```diff
@@ -0,0 +1,317 @@
+     * in "modern" mode
+     */
+    template <typename T, std::enable_if_t<std::is_same_v<T, void>, int> = 0>
+    generic_data_handle(data_handle<T> const& handle)
+        : m_container{handle.m_raw_ptr}
+        , m_type{&typeid(T*)} {
+        static_assert(std::is_same_v<T, void>);
+    }
+
+    /**
+     * @brief Wrap a data_handle<T != void> in a generic data handle.
+     */
+    template <typename T, std::enable_if_t<!std::is_same_v<T, void>, int> = 0>
+    generic_data_handle(data_handle<T> const& handle)
+        : m_offset{handle.m_offset}
+        , m_container{handle.m_container_or_raw_ptr}
+        , m_type{&typeid(T*)}
+        , m_array_dim{handle.m_array_dim}
+        , m_array_index{handle.m_array_index} {
+        static_assert(!std::is_same_v<T, void>);
+    }
+
+    /**
+     * @brief Create data_handle<T> from a generic data handle.
+     *
+     * The conversion will succeed, yielding a null data_handle<T>, if the
+     * generic data handle is null. If the generic data handle is not null then
+     * the conversion will succeed if the generic data handle actually holds a
+     * data_handle<T> or a literal T*.
+     *
+     * It might be interesting in future to explore dropping m_type in optimised
+     * builds, in which case we should aim to avoid predicating important logic
+     * on exceptions thrown by this function.
+     */
+    template <typename T>
+    explicit operator data_handle<T>() const {
+        // Either the type has to match or the generic_data_handle needs to have
+        // never been given a type
+        if (!m_type) {
+            // A (typeless / default-constructed) null generic_data_handle can
+            // be converted to any (null) data_handle<T>.
+            return {};
+        }
+        if (typeid(T*) != *m_type) {
+            throw_error(" cannot be converted to data_handle<" + cxx_demangle(typeid(T).name()) +
+                        ">");
+        }
+        if (m_offset.has_always_been_null()) {
+            // This is a data handle in backwards-compatibility mode, wrapping a
+            // raw pointer of type T*, or a null handle that has always been null (as opposed to a
+            // handle that became null). Passing do_not_search prevents the data_handle<T>
+            // constructor from trying to find the raw pointer in the NEURON data structures.
+            return {do_not_search, static_cast<T*>(m_container)};
+        }
+        if (m_offset.was_once_valid()) {
+            // This used to be a valid data handle, but it has since been
+            // invalidated. Invalid data handles never become valid again, so we
+            // can safely produce a "fully null" data_handle<T>.
+            return {};
+        }
+        if constexpr (std::is_same_v<T, void>) {
+            // This branch is needed to avoid forming references-to-void if T=void.
+            throw_error(" should not be in modern mode because it is referring to void");
```

Comment:
Agreed, deleted the line. I think the references-to-void comment is also redundant now, because we store a `T* const*` instead of a `std::vector<T>*`, and it was `std::vector<void>` that we had to avoid instantiating, but now we have the issue that `data_handle<void>` also doesn't have the array dimension/index members, so it doesn't have a 4-argument constructor.

To avoid new warnings about not all branches having return statements, I reorganised the blocks a little.